atomix/cluster/src/test/java/io/atomix/raft/cluster/impl/RaftClusterContextTest.java

```diff
@@ -376,6 +373,79 @@ void quorumWhenClusterWasSingleMember() {
     assertThat(context.getQuorumFor(RaftMemberContext::getMatchIndex)).hasValue(5L);
   }
 
+  @Test
+  void demotedMemberIsNotVoting() {
+    // given
+    final var localMember = new DefaultRaftMember(new MemberId("1"), Type.ACTIVE, Instant.now());
+
+    final var initialConfiguration =
+        new Configuration(
+            1,
+            1,
+            Instant.now().toEpochMilli(),
+            List.of(
+                localMember,
+                new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+                new DefaultRaftMember(new MemberId("3"), Type.ACTIVE, Instant.now())));
+    final var raft = raftWithStoredConfiguration(initialConfiguration);
+    final var context = new RaftClusterContext(localMember.memberId(), raft);
+
+    // when -- demote member 3 to passive
+    final var newConfiguration =
+        new Configuration(
+            2,
+            1,
+            Instant.now().toEpochMilli(),
+            List.of(
+                localMember,
+                new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+                new DefaultRaftMember(new MemberId("3"), Type.PASSIVE, Instant.now())),
+            List.of());
+    context.configure(newConfiguration);
+
+    // then -- member 3 is no longer a voting member
+    assertThat(context.getVotingMembers())
+        .containsExactly(new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()));
+  }
+
+  @Test
+  void shouldKeepMembersWithHighestType() {
+    // given - three active members
+    final var localMember = new DefaultRaftMember(new MemberId("1"), Type.ACTIVE, Instant.now());
+
+    final var initialConfiguration =
+        new Configuration(
+            1,
+            1,
+            Instant.now().toEpochMilli(),
+            List.of(
+                localMember,
+                new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+                new DefaultRaftMember(new MemberId("3"), Type.ACTIVE, Instant.now())));
+    final var raft = raftWithStoredConfiguration(initialConfiguration);
+    final var context = new RaftClusterContext(localMember.memberId(), raft);
+
+    // when -- enter joint consensus with member 3 being passive
+    final var newConfiguration =
+        new Configuration(
+            2,
+            1,
+            Instant.now().toEpochMilli(),
+            List.of(
+                localMember,
+                new DefaultRaftMember(new MemberId("2"), Type.ACTIVE, Instant.now()),
+                new DefaultRaftMember(new MemberId("3"), Type.PASSIVE, Instant.now())),
+            List.of());
+    context.configure(newConfiguration);
+
+    // then -- member 3 is no longer a voting member
```

Comment:
:wrench: comment is sort of misleading, I assume was copy pasted :smile:

Reply:
Oops yeah :+1: